contrib/src/main/scala/akka/stream/contrib/Valve.scala

```diff
@@ -0,0 +1,62 @@
+/*
+ * Copyright (C) 2016 Lightbend Inc. <http://www.lightbend.com>
+ */
+package akka.stream.contrib
+
+import akka.stream.stage.{ GraphStageLogic, GraphStageWithMaterializedValue, InHandler, OutHandler }
+import akka.stream.{ Attributes, FlowShape, Inlet, Outlet }
+
+trait ValveSwitch {
+  def open(): Unit
+  def close(): Unit
+}
+
+class Valve[A](mode: ValveMode = ValveMode.Open) extends GraphStageWithMaterializedValue[FlowShape[A, A], ValveSwitch] {
+
+  override val shape = FlowShape(Inlet[A]("valve.in"), Outlet[A]("valve.out"))
+
+  override def createLogicAndMaterializedValue(inheritedAttributes: Attributes): (GraphStageLogic, ValveSwitch) = {
+    val logic = new ValveGraphStageLogic(shape, mode)
+    (logic, logic.switch)
+  }
+
+  private class ValveGraphStageLogic(shape: Shape, var mode: ValveMode) extends GraphStageLogic(shape) with InHandler with OutHandler {
+    import shape._
+
+    var bufferedElement = Option.empty[A]
+
+    val switch = new ValveSwitch {
+      val callback = getAsyncCallback[A](push(out, _))
+
+      override def open(): Unit = {
+        mode = ValveMode.Open
+        bufferedElement.foreach(callback.invoke)
```

Comment:
You need to put the logic inside the callback instead of around the invocation of the callback.